src/Pinch/Server.hs

```diff
@@ -113,6 +126,31 @@ createServer f = ThriftServer $ \ctx req ->
             Left _     -> pure ()
         _ -> pure ()
 
+multiplex :: [(ServiceName, ThriftServer)] -> ThriftServer
+multiplex services = ThriftServer $ \ctx req -> do
+  case req of
+    RCall msg   -> go ctx req (pure . msgAppEx msg)
+    -- we cannot send the exception back, because it is a oneway call
+    -- instead let's just throw it and crash the server
+    ROneway msg -> go ctx req throwIO
+  where
+    srvMap = HM.fromList services
+
+    go :: Context -> Request a -> (ApplicationException -> IO a) -> IO a
+    go ctx req onError = do
+      let (prefix, rem) = T.span (/= ':') (messageName $ getRequestMessage req)
+      let prefix' = ServiceName prefix
+      let ctx' = addToContext prefix' ctx
+      case prefix' `HM.lookup` srvMap of
+        _ | T.null rem -> onError $ ApplicationException "Invalid method name, expecting a dot." WrongMethodName
+        Just srv -> do
+
+          reply <- unThriftServer srv ctx' $ mapRequest (\msg -> msg { messageName = T.tail rem }) req
+
+          case req of
+            ROneway _ -> pure ()
+            RCall _ -> pure $ reply { messageName = prefix <> ":" <> messageName reply }
```

Comment:
If I remember correctly, servers aren't really supposed to change the message name in the response.

In that case, we can re-use the original message name here and avoid that new string.

```suggestion
            RCall _ -> pure $ reply { messageName = messageName (getRequestMessage req) }
```

Reply:
Good point, done.